src/sql/src/query_model/rewrite/rule/simplify_outer_joins.rs

```diff
@@ -0,0 +1,186 @@
+
+        // TODO: this needs an immutable visit method
+        let _: Result<(), ()> = model.try_visit_pre_post_descendants(
+            &mut |model, box_id| {
+                let r#box = model.get_box(*box_id);
+
+                // Push the ranging quantifiers count the current node.
+                quantifier_counts.push(r#box.ranging_quantifiers().count());
+
+                // At the moment, we don't want to do any work:
+                // (1) if a match is already found, or
+                // (2) if we are in a in subtree with a non-unique ancestor chain.
+                // In the future, we might remove constraint (2).
+                // See [this comment](https://github.com/MaterializeInc/materialize/issues/10239#issuecomment-1030123237)
+                // and the preceding discussion for more details.
+                if quantifier_counts.iter().all(|count| count == &1) {
+                    // Get the ID of the quantifer that connects this box with its parent.
+                    let quantifier_id = r#box.ranging_quantifiers().nth(0).unwrap().id;
+                    // Get the propagated nulls for the output attributes of this box.
+                    let propagated_nulls = r#box.attributes.get::<PropagatedNulls>();
+
+                    // Take the rej_nulls expressed in terms of the parent's quantifiers.
+                    let mut rej_nulls_parent = rej_nulls.last().unwrap();
+
+                    // And use it to compute the rej_nulls expressed in terms of this child's quantifiers.
+                    let rej_nulls_child = rej_nulls_parent
+                        .iter()
+                        // Filter out columns that are not from this child
+                        .filter(|c| c.quantifier_id != quantifier_id)
+                        // and replace the rest with the set of nulls that
+                        // the corresponding column propagates from its input.
+                        .flat_map(|c| match propagated_nulls.get(c.position) {
+                            Some(set) => set.clone(),
+                            None => HashSet::new(),
+                        })
+                        .collect::<HashSet<_>>();
+
+                    // Push rej_nulls_child to the stack.
+                    rej_nulls.push(rej_nulls_child);
+
+                    // If this is an outer join box, we might have a potential match
+                    if let BoxType::OuterJoin(bt) = r#box.box_type {
+                        assert_eq!(r#box.input_quantifiers().count(), 2);
+
+                        let lhs = r#box.input_quantifiers().nth(0).unwrap();
+                        let rhs = r#box.input_quantifiers().nth(0).unwrap();
+
+                        let rej_lhs = rej_nulls_child.iter().any(|c| c.quantifier_id == lhs.id);
+                        let rej_rhs = rej_nulls_child.iter().any(|c| c.quantifier_id == rhs.id);
+
+                        // TODO: this classification incomplete and probably incorrect
```

Comment:
@wangandi this is most certainly wrong, but I'll let you figure out the details.